File: app/src/main/java/pl/expensive/DisplayWallets.java
package pl.expensive;

import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.util.Log;

import java.util.List;

import javax.inject.Inject;

import pl.expensive.storage.Transaction;
import pl.expensive.storage.TransactionStorage;
import pl.expensive.storage.WalletsStorage;
import rx.Observable;
import rx.Single;
import rx.Subscription;

class DisplayWallets {
    @Nullable
    private Subscription fetchWalletsSubscription;

    private final WalletsStorage walletsStorage;
    private final TransactionStorage transactionStorage;

    @Inject
    DisplayWallets(WalletsStorage walletsStorage, TransactionStorage transactionStorage) {
        this.walletsStorage = walletsStorage;
        this.transactionStorage = transactionStorage;
    }

    void runFor(final WalletsViewContract view) {
        fetchWalletsSubscription = Single.fromCallable(walletsStorage::list)
                .flatMapObservable(Observable::from)
                .filter(wallet -> !TextUtils.isEmpty(wallet.name())) // TODO: 09.01.2017 Replace with non-android version
                .map(wallet -> WalletViewModel.create(
                        wallet.name(),
                        (List<Transaction>) transactionStorage.select(wallet.uuid()),
                        wallet.currency()))
                .toList()
                .subscribe(walletViewModels -> {
                    if (walletViewModels.isEmpty()) {
                        view.showEmpty();
                    } else {
                        view.showWallets(walletViewModels);
                    }
                }, err -> {
                    Log.e("XXX", err.toString());
                    view.showFetchError();
                });
    }

    void dispose() {
        if (!isDisposed()) {
            fetchWalletsSubscription.unsubscribe();
        }
    }

    private boolean isDisposed() {
        return fetchWalletsSubscription == null || fetchWalletsSubscription.isUnsubscribed();
    }
}

File: app/src/main/java/pl/expensive/DisplayWallets.kt
package pl.expensive

import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import pl.expensive.storage.Transaction
import pl.expensive.storage.TransactionStorage
import pl.expensive.storage.Wallet
import pl.expensive.storage.WalletsStorage

internal open class DisplayWallets(private val walletsStorage: WalletsStorage,
                                   private val transactionStorage: TransactionStorage) {
    fun runFor(view: WalletsViewContract) {
        view.display { fetchWallets() }
    }

    fun WalletsViewContract.display(f: () -> Collection<Wallet>) {
        doAsync {
            val viewModels = f()
                    .map {
                        WalletViewModel(
                                it.name,
                                transactionStorage.select(it.uuid) as List<Transaction>,
                                it.currency)
                    }
            uiThread {
                if (viewModels.isEmpty()) {
                    showEmpty()
                } else {
                    showWallets(viewModels)
                }
            }
        }
    }

    private fun fetchWallets(): Collection<Wallet> = walletsStorage.list()
}
